File: server/src/krate/ktab/po.rs
use super::decimal::Decimal;
use chin_sql::ChinSqlCrud;
use std::collections::HashMap;

use chin_sql::time_type::TID;
use chin_sql::GenerateTableSchema;
use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};

use crate::model::omit_tid::OmitTID;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum KTabColumnStoreKind {
    Decimal,
    Str,
    Date,
    // Blob,
}

pub(crate) type KTabColumnViewKind = String;
#[derive(Clone, Debug, Serialize, Deserialize)]
pub(crate) struct KTabColumnMeta {
    pub(crate) idx: TID,
    pub(crate) name: String,
    pub(crate) comment: String,
    pub(crate) store_kind: KTabColumnStoreKind,
    pub(crate) view_kind: KTabColumnViewKind,
    pub(crate) required: bool,
    pub(crate) order_by: i32,
}

#[derive(Clone, Debug, Serialize, Deserialize, GenerateTableSchema)]
pub(crate) struct KTabMeta {
    #[gts_primary]
    #[gts_type = "i64"]
    pub(crate) tid: TID,

    #[gts_primary]
    #[gts_type = "i64"]
    pub(crate) omit_tid: OmitTID,

    #[gts_type = "String"]
    pub(crate) columns: HashMap<String, KTabColumnMeta>,
    pub(crate) table_name: String,
    pub(crate) table_comment: String,
    pub(crate) update_time: Option<DateTime<FixedOffset>>,
    pub(crate) real_table: bool,
}

macro_rules! type_table {
    ($sname:tt, $data_type:ty $(, #[$attr:meta])*) => {
        #[derive(Clone, Debug, Serialize, Deserialize, GenerateTableSchema, ChinSqlCrud)]
        pub(crate) struct $sname {
            #[gts_primary]
            #[gts_type = "i64"]
            pub(crate) tid: TID,

            #[gts_primary]
            #[gts_type = "i64"]
            pub(crate) omit_tid: OmitTID,

            #[gts_type = "i64"]
            pub(crate) table_id: TID,
            #[gts_type = "i64"]
            pub(crate) col_idx: TID, // actually is the insert time(unix timestamp), so it is easy for data merge
            #[gts_type = "i64"]
            pub(crate) row_idx: TID, // actually is the insert time(unix timestamp), so it is easy for data merge

            $(#[$attr])*
            pub(crate) cell_data: $data_type
        }
    }
}

type_table!(KTabCellText, String);
type_table!(KTabCellDecimal, Decimal, #[gts_type = "String"]);
type_table!(KTabCellDate, DateTime<FixedOffset>);

#[cfg(test)]
mod ktab_column_type_test {
    #[test]
    fn test() {
        use crate::krate::ktab::po::KTabColumnStoreKind;

        println!("{:?}", serde_json::to_string(&KTabColumnStoreKind::Str));
    }
}

File: web/src/krate/ktab/store/po.ts
export type KTabColumnStoreKind = "str" | "i64" | "f64" | "date" | "blob";

export type KTabColumnViewKind =
  | "string"
  | "date"
  | "decimal";

export const ktabViewToStoreKind = (
  kind: KTabColumnViewKind
): KTabColumnStoreKind => {
  switch (kind.toLowerCase()) {
    case "string":
      return "str";
    case "integer":
      return "i64";
    case "float":
      return "f64";
    case "date":
      return "date";
    case "image":
      return "str";
    case "blob":
      return "blob";
    case "bool":
      return "i64";
    default:
      throw new Error(`unable to map ${kind}`);
  }
};

export type KTabColumnMeta = {
  idx: number;
  name: string;
  comment?: string;
  store_kind: KTabColumnStoreKind;
  view_kind: KTabColumnViewKind;
  required: boolean;
  order_by: number;
};

export type KTabMeta = {
  tid: number;
  columns: Record<string, KTabColumnMeta>;
  table_name: string;
  table_comment?: string;
  create_time: Date;
  update_time?: Date;
  omit_tid?: Date;
  kspace: string;
  real_table: boolean;
};

export type KTabCell = {
  table_id: number;
  col_idx: number;
  row_idx: number;
  omit_tid?: Date;
  cell_data: KTabCellData;
}

export type KTabCellData = { Date: Date } | { Text: string };
